### PR TITLE
fix: add support for asgi 3.0 while maintaining 2.0 support.

### PR DESCRIPTION
**Pull request by jhillacre**
I encountered an issue where the ASGI app I was wrapping did not return a handler. In ASGI 3.0, this is expected. While writing tests for that change, I found other fixes or tweaks I would like to see and ended up with a test suite that included those.

This pull request is more extensive than I intended, so my apologies for that. Taking a pull request that rewrites so much can be annoying, so I would understand if you only wanted to take parts of this pull request.

This pull request makes the following changes:
* Use `asgiref` to do ASGI 3.0 and ASGI 2.0 compatibility.
* `OPTIONS` does not make sense as an allowed method in the context of CORS.
  See https://stackoverflow.com/a/68529748
  * exclude `OPTIONS` from all methods
  * exclude non-CORS `OPTIONS` requests from CORS headers.
* Do not always return safelisted headers. The inclusion of safelisted headers means skipping additional browser restrictions. See https://developer.mozilla.org/en-US/docs/Glossary/CORS-safelisted_request_header#additional_restrictions
* Return `204` for successful preflight responses and `403` for unsuccessful responses.
* Only provide `vary: Origin` header when multiple enumerated origins are valid.
* Expanded test coverage.

### CODE REVIEW

File: .coveragerc [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [run]
 branch = True
-source = asgi_middleware
+source = asgi_cors_middleware
```
Comment (mrkiura) [inline .coveragerc:3]: Good catch here. Coverage report now showing.

File: asgi_cors_middleware/middleware.py [hunk 2]
```diff
@@ -7,11 +7,15 @@
...
 import typing
 
+from asgiref.compatibility import guarantee_single_callable
 from starlette.datastructures import Headers, MutableHeaders
 
 from starlette.responses import PlainTextResponse
+from starlette.responses import Response
 
-ALL_METHODS = ("DELETE", "GET", "OPTIONS", "PATCH", "POST", "PUT")
+# OPTIONS doesn't make sense to return as an allowed method for CORS.
```
Comment (mrkiura) [inline asgi_cors_middleware/middleware.py:16]: Agreed. 💯

**Comment by mrkiura** [comment]
Looks great 🔥. Thanks for adding ASGI 3.0 support, the fixes and the comprehensive test suite.

**Review by mrkiura — APPROVED** [review]

**Comment by jhillacre** [comment]
Thanks for taking on these changes.

I saw a v0.0.2 tag, but [pypi](https://pypi.org/project/asgi-cors-middleware/) didn't get v0.0.2. I'm unsure if that was missed or if you are intentionally waiting for more changes before publishing, but I wanted to ask about it.

**Comment by mrkiura** [comment]
V0.0.2 is now on pypi. I had not gotten around to doing it. 

Thanks again for the changes. If you don't mind me asking, I am curious as to your use case and how you stumbled upon this project.

**Comment by jhillacre** [comment]
I have used `django-cors-headers` on a REST API using `djangorestframework`. I'm adding WebSockets to the same project using `django-channels` for notifications and keeping the UI current. The issue I ran into is that `django-cors-headers` is Django middleware, not ASGI middleware, where channels have their own ASGI middleware for Django authentication but not one for CORS. So I found your package on pypi looking for generic ASGI CORS middleware. There was another package I tried and can't remember which now. What worked for me with your package is that it is 100% configurable, and I wanted to share the configuration between `django-cors-headers` and `asgi-cors-middleware`. Thinking about it, I probably could wrap `asgi-cors-middleware` around everything now, but I was thinking of the compatibility of settings when I picked up this library.

**Comment by mrkiura** [comment]
Interesting. I'm glad you got some value from it. I also had a somewhat similar situation. I was building a GraphQL API with [ariadne](https://github.com/mirumee/ariadne) and Django and needed to implement GraphQL subscriptions. However, subscriptions require an async server that implements the WebSockets protocol so I created an ASGI app with `channnels`.  Only to find out that `django-cors-headers` didn't work for ASGI.

**Pull request merged**